argoverse/evaluation/eval_detection.py

```diff
@@ -0,0 +1,410 @@
+                filter_metric=self.detection_cfg.detection_metric,
+                max_detection_range=self.detection_cfg.max_detection_range,
+            )
+
+            logger.info(f"{dt_filtered.shape[0]} detections")
+            logger.info(f"{gt_filtered.shape[0]} ground truth")
+            if dt_filtered.shape[0] > 0:
+                ranked_detections, scores = rank(dt_filtered)
+                metrics = self.assign(ranked_detections, gt_filtered)
+                cls_stats[class_name] = np.hstack((metrics, scores))
+
+            cls_to_ninst[class_name] = gt_filtered.shape[0]
+        return cls_stats, cls_to_ninst
+
+    def assign(self, dts: np.ndarray, gts: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
+        """Attempt assignment of each detection to a ground truth label.
+
+        Args:
+            dts: Detections of shape (N,).
+            gts: Ground truth labels of shape (M,).
+
+        Returns:
+            metrics: Matrix of true/false positive concatenated with true positive errors (N, K + S) where K is the number
+                of true positive thresholds used for AP computation and S is the number of true positive errors.
+            scores: Corresponding scores for the true positives/false positives (N,)
+        """
+
+        # Ensure the number of boxes considered per class is at most `MAX_NUM_BOXES`.
+        if dts.shape[0] > MAX_NUM_BOXES:
+            dts = dts[:MAX_NUM_BOXES]
+
+        n_threshs = len(self.detection_cfg.affinity_threshs)
+        metrics = np.zeros((dts.shape[0], n_threshs + N_TP_ERRORS))
+
+        # Set the true positive metrics to np.nan since error is undefined on false positives.
+        metrics[:, n_threshs : n_threshs + N_TP_ERRORS] = np.nan
+        if gts.shape[0] == 0:
+            return metrics
+
+        affinity_matrix = compute_affinity_matrix(dts, gts, self.detection_cfg.affinity_fn_type)
+
+        # Get the GT label for each max-affinity GT label, detection pair.
+        gt_matches = affinity_matrix.argmax(axis=1)[np.newaxis, :]
+
+        # The affinity matrix is an N by M matrix of the detections and ground truth labels respectively.
+        # We want to take the corresponding affinity for each of the initial assignments using `gt_matches`.
+        # The following line grabs the max affinity for each detection to a ground truth label.
+        affinities = np.take_along_axis(affinity_matrix.T, gt_matches, axis=0).squeeze(0)
+
+        # Find the indices of the "first" detection assigned to each GT.
+        unique_gt_matches, unique_dt_matches = np.unique(gt_matches, return_index=True)
+        for i, thr in enumerate(self.detection_cfg.affinity_threshs):
+
+            # tp_mask may need to be defined differently with other affinities
+            tp_mask = affinities[unique_dt_matches] > -thr
+            metrics[unique_dt_matches, i] = tp_mask
+
+            if thr == self.detection_cfg.tp_thresh and np.count_nonzero(tp_mask) > 0:
```

Comment:
i might add an inline comment just mentioning here that AOE, ATE, ASE are only computed with a single matching -- the one with TP @2 m